File: mobile/src/models/production.ts
// Production Planning Models

export enum ProductionStatus {
  PLANNED = 'planned',
  IN_PROGRESS = 'in_progress',
  ON_HOLD = 'on_hold',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
}

export enum ProductionPriority {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  URGENT = 'urgent',
}

export enum ProductionType {
  BATCH = 'batch',
  CONTINUOUS = 'continuous',
  JOB_SHOP = 'job_shop',
  ASSEMBLY = 'assembly',
  CUSTOM = 'custom',
}

// Base Models
export interface ProductionPlanBase {
  title: string;
  description?: string;
  production_type: ProductionType;
  priority: ProductionPriority;
  planned_start_date?: string;
  planned_end_date?: string;
  target_quantity: number;
  unit_of_measure: string;
  production_line?: string;
  equipment_required: string[];
  materials_required: MaterialRequirement[];
  labor_requirements: LaborRequirement[];
  estimated_material_cost: number;
  estimated_labor_cost: number;
  quality_standards?: string;
  inspection_points: InspectionPoint[];
  tolerance_specs: ToleranceSpec[];
  project_id?: string;
  work_order_id?: string;
  assigned_to_id?: string;
  tags: string[];
}

export interface ProductionStepBase {
  step_number: number;
  step_name: string;
  description?: string;
  estimated_duration_minutes: number;
  equipment_required: string[];
  materials_consumed: MaterialConsumption[];
  labor_required: LaborRequirement[];
  quality_checkpoints: QualityCheckpoint[];
  inspection_required: boolean;
  depends_on_steps: number[];
  notes?: string;
}

export interface ProductionScheduleBase {
  scheduled_start: string;
  scheduled_end: string;
  resource_allocation: ResourceAllocation;
  capacity_utilization: number;
  constraints: string[];
  dependencies: string[];
}

// Create Models
export interface ProductionPlanCreate extends ProductionPlanBase {}

export interface ProductionStepCreate extends ProductionStepBase {}

export interface ProductionScheduleCreate extends ProductionScheduleBase {
  production_plan_id: string;
}

// Update Models
export interface ProductionPlanUpdate {
  title?: string;
  description?: string;
  production_type?: ProductionType;
  status?: ProductionStatus;
  priority?: ProductionPriority;
  planned_start_date?: string;
  planned_end_date?: string;
  actual_start_date?: string;
  actual_end_date?: string;
  target_quantity?: number;
  actual_quantity?: number;
  unit_of_measure?: string;
  production_line?: string;
  equipment_required?: string[];
  materials_required?: MaterialRequirement[];
  labor_requirements?: LaborRequirement[];
  estimated_material_cost?: number;
  estimated_labor_cost?: number;
  actual_material_cost?: number;
  actual_labor_cost?: number;
  quality_standards?: string;
  inspection_points?: InspectionPoint[];
  tolerance_specs?: ToleranceSpec[];
  project_id?: string;
  work_order_id?: string;
  assigned_to_id?: string;
  completion_percentage?: number;
  current_step?: string;
  notes?: ProductionNote[];
  tags?: string[];
}

export interface ProductionStepUpdate {
  step_name?: string;
  description?: string;
  estimated_duration_minutes?: number;
  actual_duration_minutes?: number;
  status?: string;
  equipment_required?: string[];
  materials_consumed?: MaterialConsumption[];
  labor_required?: LaborRequirement[];
  quality_checkpoints?: QualityCheckpoint[];
  inspection_required?: boolean;
  depends_on_steps?: number[];
  notes?: string;
}

export interface ProductionScheduleUpdate {
  scheduled_start?: string;
  scheduled_end?: string;
  resource_allocation?: ResourceAllocation;
  capacity_utilization?: number;
  constraints?: string[];
  dependencies?: string[];
  status?: string;
}

// Response Models
export interface ProductionStepResponse extends ProductionStepBase {
  id: string;
  production_plan_id: string;
  status: string;
  actual_duration_minutes: number;
  created_at: string;
  updated_at: string;
}

export interface ProductionScheduleResponse extends ProductionScheduleBase {
  id: string;
  production_plan_id: string;
  tenant_id: string;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface ProductionPlanResponse extends ProductionPlanBase {
  id: string;
  tenant_id: string;
  plan_number: string;
  status: ProductionStatus;
  actual_start_date?: string;
  actual_end_date?: string;
  estimated_duration_hours: number;
  actual_duration_hours: number;
  actual_quantity: number;
  completion_percentage: number;
  current_step?: string;
  notes: ProductionNote[];
  created_at: string;
  updated_at: string;
  production_steps: ProductionStepResponse[];
  production_schedules: ProductionScheduleResponse[];
}

// List Response Models
export interface ProductionPlansResponse {
  production_plans: ProductionPlanResponse[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface ProductionStepsResponse {
  production_steps: ProductionStepResponse[];
  total: number;
}

export interface ProductionSchedulesResponse {
  production_schedules: ProductionScheduleResponse[];
  total: number;
}

// Dashboard Models
export interface ProductionDashboardStats {
  total_plans: number;
  status_counts: Record<string, number>;
  priority_counts: Record<string, number>;
  completion_rate: number;
  completed_plans: number;
  in_progress_plans: number;
  planned_plans: number;
  on_hold_plans: number;
  cancelled_plans: number;
}

export interface ProductionDashboard {
  stats: ProductionDashboardStats;
  recent_plans: ProductionPlanResponse[];
  upcoming_deadlines: ProductionPlanResponse[];
  priority_alerts: ProductionPlanResponse[];
}

// Filter Models
export interface ProductionPlanFilters {
  status?: ProductionStatus;
  priority?: ProductionPriority;
  production_type?: ProductionType;
  project_id?: string;
  work_order_id?: string;
  assigned_to_id?: string;
  planned_start_date_from?: string;
  planned_start_date_to?: string;
  planned_end_date_from?: string;
  planned_end_date_to?: string;
  search?: string;
}

// Supporting Interfaces
export interface MaterialRequirement {
  material_id: string;
  material_name: string;
  quantity: number;
  unit: string;
  cost_per_unit: number;
  total_cost: number;
}

export interface MaterialConsumption {
  material_id: string;
  material_name: string;
  quantity_consumed: number;
  unit: string;
  cost_per_unit: number;
  total_cost: number;
}

export interface LaborRequirement {
  role: string;
  hours_required: number;
  hourly_rate: number;
  total_cost: number;
}

export interface QualityCheckpoint {
  checkpoint_name: string;
  description: string;
  acceptance_criteria: string;
  measurement_method: string;
  tolerance: string;
}

export interface ToleranceSpec {
  parameter: string;
  nominal_value: number;
  upper_tolerance: number;
  lower_tolerance: number;
  unit: string;
}

export interface InspectionPoint {
  point_name: string;
  description: string;
  location: string;
  frequency: string;
  inspector_role: string;
}

export interface ResourceAllocation {
  equipment: EquipmentAllocation[];
  labor: LaborAllocation[];
  materials: MaterialAllocation[];
}

export interface EquipmentAllocation {
  equipment_id: string;
  equipment_name: string;
  start_time: string;
  end_time: string;
  utilization_percentage: number;
}

export interface LaborAllocation {
  role: string;
  employee_id?: string;
  employee_name?: string;
  hours_allocated: number;
  start_time: string;
  end_time: string;
}

export interface MaterialAllocation {
  material_id: string;
  material_name: string;
  quantity_allocated: number;
  unit: string;
  delivery_time: string;
}

export interface ProductionNote {
  id: string;
  content: string;
  author_id: string;
  author_name: string;
  timestamp: string;
  type: 'general' | 'quality' | 'safety' | 'technical';
}

// Utility Types
export type ProductionPlanFormData = Omit<
  ProductionPlanCreate,
  | 'materials_required'
  | 'labor_requirements'
  | 'inspection_points'
  | 'tolerance_specs'
  | 'tags'
> & {
  materials_required: MaterialRequirement[];
  labor_requirements: LaborRequirement[];
  inspection_points: InspectionPoint[];
  tolerance_specs: ToleranceSpec[];
  tags: string[];
};

export type ProductionStepFormData = Omit<
  ProductionStepCreate,
  | 'equipment_required'
  | 'materials_consumed'
  | 'labor_required'
  | 'quality_checkpoints'
  | 'depends_on_steps'
> & {
  equipment_required: string[];
  materials_consumed: MaterialConsumption[];
  labor_required: LaborRequirement[];
  quality_checkpoints: QualityCheckpoint[];
  depends_on_steps: number[];
};

File: backend/src/models/production.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum

# Enums
class ProductionStatus(str, Enum):
    PLANNED = "planned"
    IN_PROGRESS = "in_progress"
    ON_HOLD = "on_hold"
    COMPLETED = "completed"
    CANCELLED = "cancelled"

class ProductionPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"

class ProductionType(str, Enum):
    BATCH = "batch"
    CONTINUOUS = "continuous"
    JOB_SHOP = "job_shop"
    ASSEMBLY = "assembly"
    CUSTOM = "custom"

# Base Models
class ProductionPlanBase(BaseModel):
    title: str = Field(..., description="Production plan title")
    description: Optional[str] = Field(None, description="Production plan description")
    production_type: ProductionType = Field(ProductionType.BATCH, description="Type of production")
    priority: ProductionPriority = Field(ProductionPriority.MEDIUM, description="Production priority")
    planned_start_date: Optional[datetime] = Field(None, description="Planned start date")
    planned_end_date: Optional[datetime] = Field(None, description="Planned end date")
    target_quantity: int = Field(0, description="Target production quantity")
    unit_of_measure: str = Field("pieces", description="Unit of measure for quantity")
    production_line: Optional[str] = Field(None, description="Production line identifier")
    equipment_required: List[str] = Field(default_factory=list, description="Required equipment")
    materials_required: List[Dict[str, Any]] = Field(default_factory=list, description="Required materials")
    labor_requirements: List[Dict[str, Any]] = Field(default_factory=list, description="Labor requirements")
    estimated_material_cost: float = Field(0.0, description="Estimated material cost")
    estimated_labor_cost: float = Field(0.0, description="Estimated labor cost")
    quality_standards: Optional[str] = Field(None, description="Quality standards")
    inspection_points: List[Dict[str, Any]] = Field(default_factory=list, description="Inspection points")
    tolerance_specs: List[Dict[str, Any]] = Field(default_factory=list, description="Tolerance specifications")
    project_id: Optional[str] = Field(None, description="Associated project ID")
    work_order_id: Optional[str] = Field(None, description="Associated work order ID")
    assigned_to_id: Optional[str] = Field(None, description="Assigned user ID")
    tags: List[str] = Field(default_factory=list, description="Tags for categorization")

class ProductionStepBase(BaseModel):
    step_number: int = Field(..., description="Step sequence number")
    step_name: str = Field(..., description="Step name")
    description: Optional[str] = Field(None, description="Step description")
    estimated_duration_minutes: int = Field(0, description="Estimated duration in minutes")
    equipment_required: List[str] = Field(default_factory=list, description="Required equipment")
    materials_consumed: List[Dict[str, Any]] = Field(default_factory=list, description="Materials consumed")
    labor_required: List[Dict[str, Any]] = Field(default_factory=list, description="Labor required")
    quality_checkpoints: List[Dict[str, Any]] = Field(default_factory=list, description="Quality checkpoints")
    inspection_required: bool = Field(False, description="Whether inspection is required")
    depends_on_steps: List[int] = Field(default_factory=list, description="Dependent step numbers")
    notes: Optional[str] = Field(None, description="Additional notes")

class ProductionScheduleBase(BaseModel):
    scheduled_start: datetime = Field(..., description="Scheduled start time")
    scheduled_end: datetime = Field(..., description="Scheduled end time")
    resource_allocation: Dict[str, Any] = Field(default_factory=dict, description="Resource allocation")
    capacity_utilization: float = Field(0.0, description="Capacity utilization percentage")
    constraints: List[str] = Field(default_factory=list, description="Scheduling constraints")
    dependencies: List[str] = Field(default_factory=list, description="Dependent production plans")

# Create Models
class ProductionPlanCreate(ProductionPlanBase):
    pass

class ProductionStepCreate(ProductionStepBase):
    pass

class ProductionScheduleCreate(ProductionScheduleBase):
    production_plan_id: str = Field(..., description="Production plan ID")

# Update Models
class ProductionPlanUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    production_type: Optional[ProductionType] = None
    status: Optional[ProductionStatus] = None
    priority: Optional[ProductionPriority] = None
    planned_start_date: Optional[datetime] = None
    planned_end_date: Optional[datetime] = None
    actual_start_date: Optional[datetime] = None
    actual_end_date: Optional[datetime] = None
    target_quantity: Optional[int] = None
    actual_quantity: Optional[int] = None
    unit_of_measure: Optional[str] = None
    production_line: Optional[str] = None
    equipment_required: Optional[List[str]] = None
    materials_required: Optional[List[Dict[str, Any]]] = None
    labor_requirements: Optional[List[Dict[str, Any]]] = None
    estimated_material_cost: Optional[float] = None
    estimated_labor_cost: Optional[float] = None
    actual_material_cost: Optional[float] = None
    actual_labor_cost: Optional[float] = None
    quality_standards: Optional[str] = None
    inspection_points: Optional[List[Dict[str, Any]]] = None
    tolerance_specs: Optional[List[Dict[str, Any]]] = None
    project_id: Optional[str] = None
    work_order_id: Optional[str] = None
    assigned_to_id: Optional[str] = None
    completion_percentage: Optional[float] = None
    current_step: Optional[str] = None
    notes: Optional[List[Dict[str, Any]]] = None
    tags: Optional[List[str]] = None

class ProductionStepUpdate(BaseModel):
    step_name: Optional[str] = None
    description: Optional[str] = None
    estimated_duration_minutes: Optional[int] = None
    actual_duration_minutes: Optional[int] = None
    status: Optional[str] = None
    equipment_required: Optional[List[str]] = None
    materials_consumed: Optional[List[Dict[str, Any]]] = None
    labor_required: Optional[List[Dict[str, Any]]] = None
    quality_checkpoints: Optional[List[Dict[str, Any]]] = None
    inspection_required: Optional[bool] = None
    depends_on_steps: Optional[List[int]] = None
    notes: Optional[str] = None

class ProductionScheduleUpdate(BaseModel):
    scheduled_start: Optional[datetime] = None
    scheduled_end: Optional[datetime] = None
    resource_allocation: Optional[Dict[str, Any]] = None
    capacity_utilization: Optional[float] = None
    constraints: Optional[List[str]] = None
    dependencies: Optional[List[str]] = None
    status: Optional[str] = None

# Response Models
class ProductionStepResponse(ProductionStepBase):
    id: str
    production_plan_id: str
    status: str
    actual_duration_minutes: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class ProductionScheduleResponse(ProductionScheduleBase):
    id: str
    production_plan_id: str
    tenant_id: str
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class ProductionPlanResponse(ProductionPlanBase):
    id: str
    tenant_id: str
    plan_number: str
    status: ProductionStatus
    actual_start_date: Optional[datetime]
    actual_end_date: Optional[datetime]
    estimated_duration_hours: float
    actual_duration_hours: float
    actual_quantity: int
    completion_percentage: float
    current_step: Optional[str]
    notes: List[Dict[str, Any]]
    created_at: datetime
    updated_at: datetime
    production_steps: List[ProductionStepResponse] = []
    production_schedules: List[ProductionScheduleResponse] = []

    class Config:
        from_attributes = True
        
    @classmethod
    def from_orm(cls, obj):
        data = {
            'id': str(obj.id),
            'tenant_id': str(obj.tenant_id),
            'plan_number': obj.plan_number,
            'title': obj.title,
            'description': obj.description,
            'production_type': obj.production_type,
            'status': obj.status,
            'priority': obj.priority,
            'planned_start_date': obj.planned_start_date,
            'planned_end_date': obj.planned_end_date,
            'actual_start_date': obj.actual_start_date,
            'actual_end_date': obj.actual_end_date,
            'target_quantity': obj.target_quantity,
            'actual_quantity': obj.actual_quantity,
            'unit_of_measure': obj.unit_of_measure,
            'production_line': obj.production_line,
            'equipment_required': obj.equipment_required or [],
            'materials_required': obj.materials_required or [],
            'labor_requirements': obj.labor_requirements or [],
            'estimated_material_cost': obj.estimated_material_cost,
            'estimated_labor_cost': obj.estimated_labor_cost,
            'actual_material_cost': obj.actual_material_cost,
            'actual_labor_cost': obj.actual_labor_cost,
            'quality_standards': obj.quality_standards,
            'inspection_points': obj.inspection_points or [],
            'tolerance_specs': obj.tolerance_specs or [],
            'project_id': str(obj.project_id) if obj.project_id else None,
            'work_order_id': str(obj.work_order_id) if obj.work_order_id else None,
            'assigned_to_id': str(obj.assigned_to_id) if obj.assigned_to_id else None,
            'estimated_duration_hours': obj.estimated_duration_hours,
            'actual_duration_hours': obj.actual_duration_hours,
            'completion_percentage': obj.completion_percentage,
            'current_step': obj.current_step,
            'notes': obj.notes or [],
            'tags': obj.tags or [],
            'created_at': obj.created_at,
            'updated_at': obj.updated_at,
            'production_steps': [],
            'production_schedules': []
        }
        return cls(**data)

# List Response Models
class ProductionPlansResponse(BaseModel):
    production_plans: List[ProductionPlanResponse]
    total: int
    page: int
    limit: int
    total_pages: int

class ProductionStepsResponse(BaseModel):
    production_steps: List[ProductionStepResponse]
    total: int

class ProductionSchedulesResponse(BaseModel):
    production_schedules: List[ProductionScheduleResponse]
    total: int

# Dashboard Models
class ProductionDashboardStats(BaseModel):
    total_plans: int
    status_counts: Dict[str, int]
    priority_counts: Dict[str, int]
    completion_rate: float
    completed_plans: int
    in_progress_plans: int
    planned_plans: int
    on_hold_plans: int
    cancelled_plans: int

class ProductionDashboard(BaseModel):
    stats: ProductionDashboardStats
    recent_plans: List[ProductionPlanResponse]
    upcoming_deadlines: List[ProductionPlanResponse]
    priority_alerts: List[ProductionPlanResponse]

# Filter Models
class ProductionPlanFilters(BaseModel):
    status: Optional[ProductionStatus] = None
    priority: Optional[ProductionPriority] = None
    production_type: Optional[ProductionType] = None
    project_id: Optional[str] = None
    work_order_id: Optional[str] = None
    assigned_to_id: Optional[str] = None
    planned_start_date_from: Optional[datetime] = None
    planned_start_date_to: Optional[datetime] = None
    planned_end_date_from: Optional[datetime] = None
    planned_end_date_to: Optional[datetime] = None
    search: Optional[str] = None
